### DISCUSSION TITLE
MKS Robin nano v.3 using Marlin color UI ...

### DISCUSSION THREAD

**Discussion by themacboy** [luc-github/ESP3D#631] [Questions]
Good morning,

Im currently Im using MKS Robin nano v.2 and v.3 boards un my printers. After discovering the ESP3d project I am excited, but I have some issues.

The most important for me is that Im not usinf the MKS UI (LVGL, it is horrible), Im in love with Marlin Color UI and touchscreen.

Aftger a few attemps to installing ESP3D I get succes and all looks fine, but unfotunatelly I don't have acces to upload files in the SD card or the new USB drive support, please check pictures:

![imatge](https://user-images.githubusercontent.com/8063711/117531766-85f8a000-afe4-11eb-8e22-266cf57b6272.png)

Yo can notice that UPLOAD button is missing and I can't send files to print (print button missing too in USB Mode)

Now check SD card:

![imatge](https://user-images.githubusercontent.com/8063711/117531793-af193080-afe4-11eb-82a2-083ebef0eeca.png)

As you can notice I can't UPLOAD (missing button) but curiously I can print current files inside the SDS Card.

I have set, in Marlin, the second serial port in 1. But not sure if I need to configure something else.

Btw, someone in MKS forums has told me that ESP3D support in Marlin must be dissabled:

![imatge](https://user-images.githubusercontent.com/8063711/117531847-ee478180-afe4-11eb-802f-617b6db78975.png)

Any idea what im doing wrong?

**Comment by luc-github** [comment]
let me clarify MKS and Marlin situation and esp3d:
1 - Marlin by default use serial communication, there is not fast upload but M28/M29 protocol which is super slow 
This slow like upload is implemented in ESP3D 2.X, this protocol is no reliable, it is slow (0.4KB/S),  lot of people complained about it and blame ESP3D - so In 3.0 - I just did not implemented because it is waste of time and I am already overloaded

2 - MKS implemented a fast upload and until recently the protocol was not open, when it was opened I have implemented it in ESP3D 3.0 with support of MKS devt team,  it is limited to MKS boards and MKS wifi module because use  2 additionals pins on top of Rx/Tx and use very high bauderate that does not tolerate any error and so any EMI noise on communication. the speed is 110KB on my side but it may depend of the SD card also
This protocol also rely on MKS implementation in Marlin : https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/blob/master/Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp

So:
 if you do not use MKS protocol (in Marlin and ESP3D) -> case 1 -> no sd upload in ESP3D 3.0  but only in 2.1
 if you use MKS protocol (in Marlin and ESP3D)  ->case 2 > fast upload in ESP3D 3.0 but not in 2.1
 if you mixed serial and mks protocol -> no communication in any ESP3D version
 
 again there is a status in github that I upldate when there are update:
 https://github.com/luc-github/ESP3D/discussions/540#discussioncomment-191759
 
 And the FAQ are also describing the situation: https://github.com/luc-github/ESP3D/discussions/620
 
@themacboy  in your case you are in pure serial communication - so no SD upload in 3.0 but only in 2.1 for the moment

**Comment by themacboy** [comment]
> let me clarify MKS and Marlin situation and esp3d:
> 1 - Marlin by default use serial communication, there is not fast upload but M28/M29 protocol which is super slow
> This slow like upload is implemented in ESP3D 2.X, this protocol is no reliable, it is slow (0.4KB/S), lot of people complained about it and blame ESP3D - so In 3.0 - I just did not implemented because it is waste of time and I am already overloaded
> 
> 2 - MKS implemented a fast upload and until recently the protocol was not open, when it was opened I have implemented it in ESP3D 3.0 with support of MKS devt team, it is limited to MKS boards and MKS wifi module because use 2 additionals pins on top of Rx/Tx and use very high bauderate that does not tolerate any error and so any EMI noise on communication. the speed is 110KB on my side but it may depend of the SD card also
> This protocol also rely on MKS implementation in Marlin : https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/blob/master/Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp
> 
> So:
> if you do not use MKS protocol (in Marlin and ESP3D) -> case 1 -> no sd upload in ESP3D 3.0 but only in 2.1
> if you use MKS protocol (in Marlin and ESP3D) ->case 2 > fast upload in ESP3D 3.0 but not in 2.1
> if you mixed serial and mks protocol -> no communication in any ESP3D version
> 
> again there is a status in github that I upldate when there are update:
> [#540 (comment)](https://github.com/luc-github/ESP3D/discussions/540#discussioncomment-191759)
> 
> And the FAQ are also describing the situation: #620
> 
> @themacboy in your case you are in pure serial communication - so no SD upload in 3.0 but only in 2.1 for the moment

Thx for you explanation, now I understand the situation (sorry my english is not really good and sometimes I miss understoiod a lot of comments).

Then I will wait, for me currently is enought with control of the printer using web UI and the terminal.

Thx for you effort.

**Reply by liuhui11155** [reply]
How to change the mks protocol on marlin

**Reply by luc-github** [reply]
what do you mean `change the mks protocol in marlin` ?

**Reply by liuhui11155** [reply]
I will not modify the code in wifi_module.cpp

**Reply by luc-github** [reply]
sorry I still do not understand the question

**Reply by liuhui11155** [reply]
Sorry, I don’t know English because I just want to know how to set it to "use MKS protocol (in Marlin and ESP3D)"

**Reply by luc-github** [reply]
do you have mks robin board ?  you just have to enable it in Marlin
for ESP3D 3.0 in configuration.h  enable MKS SERIAL protocol  and read https://github.com/luc-github/ESP3D/discussions/620

**Comment by luc-github** [comment]
Thank you I move question to discussion then
